### ISSUE TITLE
Ubuntu downloads are insecure and offer poor security usability

### ISSUE THREAD

**Issue by eligrey** [canonical-websites/www.ubuntu.com#4394]
(This is a continuation of #822, which I do not believe was fixed)

Issue #822 was closed with [this comment](https://github.com/canonical-websites/www.ubuntu.com/issues/822#issuecomment-360822489):

> This issues is solved in the strictest sense, in that the guide at https://tutorials.ubuntu.com/tutorial/tutorial-how-to-verify-ubuntu can be trusted, as it is now served over HTTPS, and therefore the GPG fingerprints found on step 2 can also be trusted, allowing you to verify your download.

The principal issue is not solved. Giving users a how-to guide isn't a solution for users that will immediately use the downloaded ISO. There are only two solutions, and neither of them have been implemented.

1. (Very easy solution) Serve releases.ubuntu.com over HTTPS. If I recall correctly this was already done in the past. Why has this regressed?
2. (Harder solution) Automatically verify the GPG signature using a client-side GPG implementation hosted on the already-secured <https://www.ubuntu.com/download/desktop>. Add CORS headers to releases.ubuntu.com so that www.ubuntu.com can access the file contents to perform GPG verification automatically. Only save the file for the user if it passes the GPG signature check. If it doesn't pass the signature check, display an error message to the user saying that the download could not be verified and may be corrupted.

Consider [this recent tweet by Matt Linton](https://twitter.com/0xMatt/status/1062782977212583936) when you think about security usability design:

> Security can only be accomplished by writing secure software and making security easy.
> 
> You cannot lay it on the feet of the users - it has to be the tools.

**Comment by nottrobin** [comment]
Sorry for the delay in responding to this @eligrey. Here's the summary of the state of Ubuntu download security, as I understand it. I

This discussion around the issue of providing Ubuntu downloads securely, possibly over HTTPS, has been going on at least since I joined Canonical in March 2014. Here are examples of previous discussions:

- 2014-08-21: [Ubuntu ISOs downloaded insecurely, over HTTP rather than HTTPS](https://bugs.launchpad.net/ubuntu/+bug/1359836)
- 2015-05-12: [Ubuntu downloads are insecure](https://bugs.launchpad.net/ubuntu-website-content/+bug/1454247)
- 2016-08-30: [Ubuntu is provided insecurely by default](https://github.com/canonical-websites/www.ubuntu.com/issues/822) (as you know)
- 2017-10-19: [releases.ubuntu.com should be available over https](https://bugs.launchpad.net/ubuntu/+bug/1724899)

There have also been emails sent directly to us, and much internal discussion about this topic.

The first thing to say is that, as noted in https://github.com/canonical-websites/www.ubuntu.com/issues/822, it is technically possible to verify the contents of Ubuntu ISO downloads through purely trusted channels. However I do completely agree that this is not sufficient. For security solutions to be effective they have to be trivial for everyone to use, and GPG verification is a very long way from satisfying that goal.

> 1. (Very easy solution) Serve releases.ubuntu.com over HTTPS. If I recall correctly this was already done in the past. Why has this regressed?

I do agree that Serving the file over HTTPS would go a long way towards allowing the user to trust the delivered download content without requiring any expertise on their part. However, unfortunately providing this is not "very easy". I am not aware of releases.ubuntu.com ever supporting HTTPS. If you have any evidence that it was ever enabled, I'd be very curious to see it.

While this team manages www.ubuntu.com itself, our download pages forward users to a mirror in their region from [the mirrors list](https://launchpad.net/ubuntu/+cdmirrors-rss). These mirrors are typically not controlled by Canonical but instead contributed by the community. Some of them, but not most of them, support HTTPS, although our mirrors list doesn't appear to even link to the HTTPS version of mirrors where it is available.

The only ISO hosts controlled by Canonical are http://releases.ubuntu.com and http://cdimage.ubuntu.com. I believe the former is the one actually included in the mirrors list. Neither of these domains are controlled by this team, but instead by the Information Services team. We have broached the topic of enabling HTTPS on these domains multiple times, and there is an ongoing discussion about it at the moment. They are currently working out the implications both of providing a secure host when most of the mirror network is not secure, and of the computational implications of serving massive downloads at huge scale over HTTPS. Since these are not the expertise of our team, we defer to them on those practicalities.

If you want to continue that discussion, about either securing release.ubuntu.com or questions about the mirror list, I'd suggest filing a bug over at https://bugs.launchpad.net/ubuntu-cdimage. Since we've already broached it with them and the discussion is ongoing, there's little more my team can do from our end.

> 2. (Harder solution) Automatically verify the GPG signature using a client-side GPG implementation hosted on the already-secured https://www.ubuntu.com/download/desktop. Add CORS headers to releases.ubuntu.com so that www.ubuntu.com can access the file contents to perform GPG verification automatically. Only save the file for the user if it passes the GPG signature check. If it doesn't pass the signature check, display an error message to the user saying that the download could not be verified and may be corrupted.

I do like the idea of exploring ways to ease the user experience for verifying downloads purely within www.ubuntu.com. However, I'm afraid I don't see how this solution would meaningfully improve the user's security. If the www.ubuntu.com server-side application were to verify the download, it would not prove anything about the download that the user will eventually get. The attack vector here is a [Man In The Middle](https://en.wikipedia.org/wiki/Man-in-the-middle_attack) between the user and the releases.ubuntu.com server, not between www.ubuntu.com and releases.ubuntu.com. The only way for www.ubuntu.com to help in that conversation is for it to actually download and re-serve the file directly from www.ubuntu.com over HTTPS, which we can't do because www.ubuntu.com is not set up to serve these huge files at this scale.

One thing we could do, however, is to provide the instructions to verify the GPG signatures on the command-line directly on the "thank you" page after triggering the download. I've filed https://github.com/canonical-websites/www.ubuntu.com/issues/4430 to continue that discussion.

I'm closing this issue now, because the principal discussion should be continued on https://bugs.launchpad.net/ubuntu-cdimage instead, and #4430 covers the only other possibiility.

**Comment by eligrey** [comment]
> If the www.ubuntu.com server-side application were to verify the download, it would not prove anything about the download that the user will eventually get

I am not referring to your server-side serving the content. I was referring to implementing client-side JavaScript on www.ubuntu.com that verifies content served with permissive [CORS headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS).

**Comment by nottrobin** [comment]
Oh I get that now. Okay, that *could* help initially. But if a MITM were aware that we do that, it wouldn't be very difficult for them to deliberately serve the correct content to our client side script to pass the test and still ultimately serve a hacked download to the user. Since this MITM would have been created specifically for this purpose, one can assume they would go to the effort of figuring that out. So it doesn't seem worth the effort of implementing it.

I have, however, proposed another issue for one more possibility: https://github.com/canonical-websites/www.ubuntu.com/issues/4431

**Comment by eligrey** [comment]
> Okay, that could help initially. But if a MITM were aware that we do that, it wouldn't be very difficult for them to deliberately serve the correct content to our client side script to pass the test and still ultimately serve a hacked download to the user

https://www.ubuntu.com/download/desktop is already secure against MITM as it uses HTTPS. This allows <https://www.ubuntu.com> to safely attest trust in third party unsecured content using a client-side GPG implementation (which of course would also be served over HTTPS on <https://www.ubuntu.com>.

**Comment by nottrobin** [comment]
If we were going to do it we would do it ourselves, as we'd obviously have to maintain it. I consider 3000 lines plus those libraries fairly heavy weight. But more generally, reimplementing core browser functionality isn't something I'm a big fan of. I think the front end team might agree with me.

I doubt asking each of our mirrors to add those CORS headers is practical, do it could likely only ever be a solution for releases.ubuntu.com, which is problematic.

Not to mention that many many people don't go through www.ubuntu.com to get the download. Many go to mirrors directly, use CLI etc.

If you like, you can create a new issue about this suggestion specifically and see what the rest of the team thinks. But I really think there's much more mileage in trying to tease out the HTTPS routes, which solve the issue in a much more standardised and wider sense, without a large maintenance burden.

**Comment by nottrobin** [comment]
> Solution one provides less security than solution two as checksums are also requested from the same mirror instead of https://www.ubuntu.com. Solution two prevents rogue mirrors from serving certain users (e.g. political dissidents) malicious versions of Ubuntu.

Well that's a good point. But asking mirrors to add CORS headers for our scripts is unlikely to go down terribly well. Maybe worth checking though.

**Comment by eligrey** [comment]
@nottrobin Sorry, that tweet was from before you sent this comment. I am deleting it as you are clearly open to the client-side solution if I were to submit a PR with the code.

It appears that you responded to the tweet but I cannot view your response.

**Comment by nottrobin** [comment]
Please do file your suggestion as another issue about that specific topic, I'll make sure to see what others think and get back to you.

I'm going to keep you blocked on Twitter. I initially thought I could discuss with you amicably, but I can really do without your attitude in my personal life. I'm sick, and going to sleep.

**Comment by eligrey** [comment]
@nottrobin You should also look into [Signed HTTP Exchanges](https://developers.google.com/web/updates/2018/11/signed-exchanges). It's basically solution #2 except there's no extra header necessary and the client-side validation code is already built-in to the browser.

**Comment by dessant** [comment]
What is the roadblock for serving releases.ubuntu.com over HTTPS besides HTTP? It would be a great start to at least serve HTTPS links to users, while the legacy infrastructure could still use HTTP.

**Comment by nottrobin** [comment]
@dessant I don't know the definitive answer to that question as this team doesn't manage releases.ubuntu.com (may be there is a significant computational overhead to encrypting such large payloads?). If you ask the question over at https://bugs.launchpad.net/ubuntu-cdimage they may be able to answer it more effectively.

However, it is possible to provide an HTTPS link to an Ubuntu image using some of [the community mirrors](https://launchpad.net/ubuntu/+cdmirrors-rss) - although they are all listed as `http://` in the mirrors file, many (most?)  of them support HTTPS, e.g. from just the very top of the file (at the time of writing):

- https://free.nchc.org.tw/ubuntu-cd/
- https://ftp.halifax.rwth-aachen.de/ubuntu-releases/
- https://mirror.init7.net/ubuntu-releases/
- https://mirror.math.princeton.edu/pub/ubuntu-iso/
- https://mirrors.layeronline.com/ubuntu-releases/
- https://mirrors.xtom.com/ubuntu-releases/
- https://es-mirrors.evowise.com/ubuntu-releases/
- https://ftp.rediris.es/sites/releases.ubuntu.com/
- https://ftp-stud.hs-esslingen.de/pub/Mirrors/releases.ubuntu.com/
- https://ftp.belnet.be/ubuntu.com/ubuntu/releases/
- https://ftp.heanet.ie/pub/ubuntu-releases/
- https://ftp.icm.edu.pl/pub/Linux/ubuntu-releases/
- https://ftp.nluug.nl/os/Linux/distr/ubuntu-releases/
- https://ftp.yz.yamagata-u.ac.jp/pub/linux/ubuntu/releases/
